stefaniaPopescu26/assignment3/Dictionary.java

```diff
@@ -36,21 +58,70 @@ public void readData(String filename) {
         }
     }
 
-    public boolean isWord(String word) {
-        return containsKey(word);
+    private void add(String word) {
+        if (word.isEmpty()) {
+            return;
+        }
+
+        TrieNode node = root;
+
+         do {
+            char letter = word.charAt(0);
+
+            if (!node.children.containsKey(letter)) {
+                node.children.put(letter, new TrieNode());
+            }
+
+            node = node.children.get(letter);
+            word = word.substring(1);
+        } while (!word.isEmpty());
+
+         node.isWord = true;
     }
 
     public boolean isPrefix(String prefix) {
-        if (prefix == null)
+        if (prefix.isEmpty()) {
             return false;
+        }
 
-        Set<String> keySet = keySet();
+        TrieNode node = root;
 
-        for (String str : keySet) {
-            if (str.startsWith(prefix))
-                return true;
+        do {
+            char letter = prefix.charAt(0);
+
+            if (!node.children.containsKey(letter)) {
+                return false;
+            }
+
+            node = node.children.get(letter);
+            prefix = prefix.substring(1);
+        } while (!prefix.isEmpty());
+
+        return true;
+    }
+
+    public boolean isWord(String word) {
```

Comment:
isWord and isPrefix you write are essentially the same with one small difference, can you maybe use a common helper function and avoid duplicating code?